pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -79,7 +80,19 @@ func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler)
 	case true:
 		// Check topic ID
 		if _, ok := s[id]; ok {
-			return ErrAlreadySubscribed
+			// Unlocking, so Unsubscribe() can access ps.subscriptions
+			ps.mu.Unlock()
+			if err := ps.Unsubscribe(id, topic); err != nil {
+				ps.mu.Lock()
```

Comment:
This lock happens only to prevent blocking in the deferred `unlock` call. Consider this:
```go
	ps.mu.Lock()
	// Check topic
	s, ok := ps.subscriptions[topic]
	switch ok {
	case true:
		// Check topic ID
		if _, ok := s[id]; ok {
			// Unlocking, so that Unsubscribe() can access ps.subscriptions
			ps.mu.Unlock()
			if err := ps.Unsubscribe(id, topic); err != nil {
				return err
			}

			ps.mu.Lock()
			s, ok = ps.subscriptions[topic]
			if !ok {
				s = make(map[string]subscription)
				ps.subscriptions[topic] = s
			}
		}
	default:
		s = make(map[string]subscription)
		ps.subscriptions[topic] = s
	}
	defer ps.mu.Unlock()
```
It looks more dangerous and complex, but should be correct and not cause any deadlocks - and you improve performance. Talking about performance - this approach holds `mu.Lock` while connecting to the message broker which is suboptimal - sometimes connecting to the broker includes opening a TCP connection, allocating system resources, even filesystem (for QoS2), and can take a lot of time. In the future, we definitely should think of possible improvements here.